extra/tensors/tensors.factor

```diff
@@ -0,0 +1,257 @@
+! Copyright (C) 2019 HMC Clinic.
+! See http://factorcode.org/license.txt for BSD license.
+USING: accessors alien.c-types alien.data arrays
+concurrency.combinators grouping kernel locals math.functions
+math.ranges math multi-methods quotations sequences sequences.private
+specialized-arrays tensors.tensor-slice typed ;
+QUALIFIED-WITH: alien.c-types c
+SPECIALIZED-ARRAY: c:float
+IN: tensors
+
+! Tensor class definition
+TUPLE: tensor
+    { shape array }
+    { vec float-array } ;
+
+! Errors
+ERROR: shape-mismatch-error shape1 shape2 ;
+
+<PRIVATE
+
+! Construct a tensor of zeros
+: <tensor> ( shape seq -- tensor )
+    tensor boa ;
+
+: >float-array ( seq -- float-array )
+    [ ] float-array{ } map-as ;
+
+:: repetition ( shape const -- tensor )
+    ! Push the shape
+    shape
+    ! Push the float array
+    shape product const <repetition>
+    >float-array
+    ! Create the tensor
+    <tensor> ;
+
+PRIVATE>
+
+! Construct a tensor of zeros
+: zeros ( shape -- tensor )
+    0 repetition ;
+
+! Construct a tensor of ones
+: ones ( shape -- tensor )
+    1 repetition ;
+
+! Construct a one-dimensional tensor with values start, start+step,
+! ..., stop (inclusive)
+: arange ( a b step -- tensor )
+    <range> [ length 1array ] keep [ ] float-array{ } map-as <tensor> ;
+
+! Construct a tensors with vec { 0 1 2 ... } and reshape to the desired shape
+: naturals ( shape -- tensor )
+    [ ] [ product [0,b) >float-array ] bi <tensor> ;
+
+! Reshape the tensor to conform to the new shape
+: reshape ( seq tensor -- tensor )
+    2dup shape>> [ product ] bi@ =
+    [ swap >>shape ]
+    [ shape>> swap shape-mismatch-error ] if ;
+
+! Flatten the tensor so that it is only one-dimensional
+: flatten ( tensor -- tensor )
+    dup shape>>
+    product { } 1sequence >>shape ;
+
+! Outputs the shape of a tensor
+: shape ( tensor -- shape )
```

Comment:
If the accessor is sufficient (meaning this is just a simple alias), we should probably just use that for now.